piqueserver/config.py

```diff
@@ -0,0 +1,145 @@
+# Copyright (c) 2017 Piqueserver development team
+
+# This file is part of piqueserver.
+
+# piqueserver is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+
+# piqueserver is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License
+# along with piqueserver.  If not, see <http://www.gnu.org/licenses/>.
+
+import json
+import toml
+
+
+DEFAULT_STYLE = 'TOML'
+TOML_STYLE = 'TOML'
+JSON_STYLE = 'JSON'
+
+
+class ConfigException(Exception):
+    pass
+
+
+class ConfigStore():
+    '''
+    configuration store that manages global configuration
+    '''
+
+    def __init__(self):
+        self._raw_config = {}
+        self.options = {}
+
+    def validate_all(self):
+        for option in self.options.values():
+            option.validate(option.get())
+
+    def clear_config(self):
+        self._raw_config = {}
+
+    def load_config(self, config_file, style=DEFAULT_STYLE):
+        self.clear_config()
+        self.update_config(config_file, style)
+
+    def update_config(self, config_file, style=DEFAULT_STYLE):
+        if style == TOML_STYLE:
+            self._raw_config.update(toml.load(open(config_file)))
```

Comment:
I don't think we should be opening this ourselves... Instead, I think we should just pass in a file-like object. Also, this can potentially leak file descriptors as the file is never closed.

Reply:
Yep good idea. Btw this shouldn't leak anything though since the garbage collector will close the files on cleanup.

Reply:
Yes, it will, in current versions of the CPython interpreter. But this is an implementation detail and can not be relied upon. In the pypy interpreter this will leak.